XFree86-3.3/xc/programs/xlsfonts/xlsfonts.man
.\" $XConsortium: xlsfonts.man /main/21 1996/12/09 17:09:56 kaleb $
.\" Copyright (c) 1988  X Consortium
.\" 
.\" Permission is hereby granted, free of charge, to any person obtaining
.\" a copy of this software and associated documentation files (the
.\" "Software"), to deal in the Software without restriction, including
.\" without limitation the rights to use, copy, modify, merge, publish,
.\" distribute, sublicense, and/or sell copies of the Software, and to
.\" permit persons to whom the Software is furnished to do so, subject to
.\" the following conditions:
.\" 
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR
.\" OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
.\" ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\" 
.\" Except as contained in this notice, the name of the X Consortium shall
.\" not be used in advertising or otherwise to promote the sale, use or
.\" other dealings in this Software without prior written authorization
.\" from the X Consortium.
.TH XLSFONTS 1 "Release 6.3" "X Version 11"
.SH NAME
xlsfonts \- server font list displayer for X
.SH SYNOPSIS
.B xlsfonts
[\-options ...] [\-fn pattern]
.SH DESCRIPTION
.I Xlsfonts
lists the fonts that match the given \fIpattern\fP.
The wildcard character "*" may be used to match any sequence of
characters (including none), and "?" to match any single character.
If no pattern is given, "*" is assumed.
.PP
The "*" and "?" characters must be quoted to prevent them from
being expanded by the shell.
.SH "OPTIONS"
.PP
.TP 8
.B \-display \fIhost\fP:\fIdpy\fP
This option specifies the X server to contact.
.PP
.TP 8
.B \-l
Lists some attributes of the font on one line in addition to its name.
.TP 8
.B \-ll
Lists font properties in addition to \fB\-l\fP output.
.TP 8
.B \-lll
Lists character metrics in addition to \fB\-ll\fP output.
.TP 8
.B \-m
This option indicates that long listings should also print the minimum and
maximum bounds of each font.
.TP 8
.B \-C
This option indicates that listings should use multiple columns.  This is the
same as \fB\-n 0\fP.
.TP 8
.B \-1
This option indicates that listings should use a single column.  This is the
same as \fB\-n 1\fP.
.TP 8
.B \-w \fIwidth\fP
This option specifies the width in characters that should be used in 
figuring out how many columns to print.  The default is 79.
.TP 8
.B \-n \fIcolumns\fP
This option specifies the number of columns to use in displaying the output.
By default, it will attempt to fit as many columns of font names into the 
number of character specified by \fB\-w \fIwidth\fR.
.TP 8
.B \-u
This option indicates that the output should be left unsorted.
.TP 8
.B \-o
This option indicates that \fIxlsfonts\fP should do an \fBOpenFont\fP (and
\fBQueryFont\fP, if appropriate) rather than a \fBListFonts\fP.  This is
useful if \fBListFonts\fP or \fBListFontsWithInfo\fP fail to list a known
font (as is the case with some scaled font systems).
.TP 8
.B \-fn \fIpattern\fP
This option specifies the font name pattern to match.
.PP
.SH "SEE ALSO"
X(1), Xserver(1), xset(1), xfd(1),
.I "X Logical Font Description Conventions"
.SH ENVIRONMENT
.TP 8
.B DISPLAY
to get the default host and display to use.
.SH BUGS
Doing ``xlsfonts -l'' can tie up your server for a very long time.
This is really a bug with single-threaded non-preemptable servers, not with
this program.
.SH AUTHOR
Mark Lillibridge, MIT Project Athena; Jim Fulton, MIT X Consortium;
Phil Karlton, SGI


ace/ChangeLog
Fri Oct 05 05:56:50 2001  Balachandran  <bala@cs.wustl.edu>

        * ACE version 5.2 released.

Thu Oct  4 17:38:21 2001  Steve Huston  <shuston@riverace.com>

        * ace/Synch.h: Added some comments to explain the rationale for
          the different return values from the blocking and timed
          versions of ACE_Null_Semaphore::acquire().  Thanks to Doug
          Schmidt for explaining this.

Thu Oct  4 17:27:39 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Handler.cpp (recv_log_record): Insure that
          the payload message block is properly aligned for any CDR
          extraction. Many thanks to Bala for noticing this problem and
          sending code for the fix.

Thu Oct  4 17:08:39 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Process_Per_Connection_Logging_Server.cpp:
          In prepare(), call ACE_Process_Options::command_line() rather
          than ACE_Process_Options::process_name() to set the program
          path for the child process.

Thu Oct  4 13:18:37 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Handler.cpp: Use a ACE_Message_Block,
          rather than hijacking a CDR block, to read the payload into.

Thu Oct  4 10:14:34 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Added Donald Acton to the Hall of Fame.

Thu Oct  4 06:39:20 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/config-sunos5.5.h: Added ACE_LACKS_MKSTEMP as SunOS 5.5 does
	  not support mkstemp (). 
	
	* ace/config-sunos5.6.h: Undefined ACE_LACKS_MKSTEMP as Sun 5.6
	  and above support it. 

	  Thanks to Don Hinton <dhinton@swan.fr> for reporting this
	  problem and Chad Elliott for testing this out on a Sun OS 5.5
	  box. 

Wed Oct  3 16:37:41 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/CDR_Stream.cpp (read_1): Reverted this change "Wed Oct  3
	  15:05:43 2001  Jeff Parsons <parsons@cs.wustl.edu>".  This
	  change is alone not enough to fix the problem that Brad
	  reported. The sample program for the text needs to be fixed to
	  get it working right.

Wed Oct  3 15:05:43 2001  Jeff Parsons <parsons@cs.wustl.edu>

        * ace/CDR_Stream.cpp (read_1);

          Reverted changed made in 

          Fri Aug 17 16:56:16 2001  Jeff Parsons <parsons@cs.wustl.edu>

          to read_1(). A similar change made in several other methods
          is ok, but in this case bytes can be copied into the
          input CDR stream without changing wr_ptr, so the check will
          fail unnecessarily. Thanks to Brad Hoskins <BHoskins@slo.newport.com>
          for pointing this out.

Wed Oct  2 12:18:37 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Server.cpp: Fixed spelling error.
        * examples/C++NPv1/Logging_Handler.cpp: Comment length adjustment.

Tue Oct  2 21:52:31 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Client.cpp: Reverted this:
          Mon Oct  1 18:27:41 2001  Steve Huston  <shuston@riverace.com>

Tue Oct  2 19:52:57 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* Makefile: Removed some of the files that no longer exist in the
	  repository. The inclusion of these files in the Makefile added
	  noise during the beta cutting process.

Tue Oct 02 15:34:49 2001  Carlos O'Ryan  <coryan@uci.edu>

	* etc/tao_cosevent.doxygen:
	* etc/tao_cosnotification.doxygen:
	  Fixed typos in the tag file names.

Tue Oct  2 15:52:27 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Added Sunny Leung to the Hall of Fame.

Tue Oct  2 15:33:36 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Removed one instance of Peter Phillips
	  <pphillip@opentext.com> from the Hall of Fame.

Tue Oct  2 15:00:37 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/CDR_Stream.cpp (clone_from): Copied the GIOP major and minor
	  versions from the incoming CDR stream. Thanks to Sunny Leung 
	  <sleung@opentext.com> for reporting this. This closes bugzilla
	  #1043.  

Tue Oct  2 09:45:39 2001  Steve Huston  <shuston@riverace.com>

        * tests/libDLL_Test.icc:
        * tests/libService_Config_DLL.icc: Add link(exportAll) to the link
          options to export the entrypoints needed for the tests.

Mon Oct  1 20:04:14 2001  Priyanka Gontla  <pgontla@ece.uci.edu>

	* THANKS: 
	  Added Tushar Nair to the THANKS file.

Mon Oct  1 18:06:48 2001  Priyanka Gontla  <pgontla@ece.uci.edu>

	* ace/Object_Manager.cpp (ACE_UnhandledExceptionFilter):
	  Removed the un-needed #if 0's. Left them before as a safety
	  measure (in case something doesnt work) and forgot to remove
	  them later.

	* ace/Log_Msg.cpp (instance):
	  Fixed the seg faults that are raised when ace is compiled with
	  ACE_NTRACE=0 on SunOS. Added a check before locking and
	  releasing a lock.

	  Thanks to Tushar Nair <ntushar@yahoo.com> for reporting the
	  problem and for providing a fix.

Mon Oct  1 18:27:41 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Client.cpp: Remove the third (optional)
          arg from getline(). Using a char literal without further
          qualification causes confusion in Unicode builds.

Mon Oct  1 14:44:13 2001  Ossama Othman  <ossama@uci.edu>

	* ace/SSL/SSL_Context.cpp (ssl_library_init):

	  Initialize the locking callbacks in the crypto library before
	  initializing the OpenSSL library.  OpenSSL may require the
	  locking callbacks to be initialized before its own
	  initialization.

	* ace/SSL/SSL_Context.cpp (ssl_library_fini):

	  Perform crypto library resource cleanup before reclaiming the
	  locks used by the locking callbacks.  Crypto library resource
	  cleanup makes use of the locking callbacks.

Mon Oct  1 15:19:11 2001  Douglas C. Schmidt  <schmidt@macarena.cs.wustl.edu>


XFree86-3.3/xc/programs/xlsfonts/xlsfonts.c
/*
 * $XConsortium: xlsfonts.c,v 1.34 94/04/17 20:24:37 rws Exp $
 *
 * 
Copyright (c) 1989  X Consortium

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of the X Consortium shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from the X Consortium.
 * */

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xos.h>
#include <stdio.h>
#include "dsimple.h"

#define N_START 1000  /* Maximum # of fonts to start with */

int max_output_line_width = 79;
int output_line_padding = 3;
int columns = 0;

#define L_SHORT 0
#define L_MEDIUM 1
#define L_LONG 2
#define L_VERYLONG 3

Bool	sort_output = True;
Bool	open_instead_of_list = False;
int	long_list = L_SHORT;
int	nnames = N_START;
int	font_cnt;
int	min_max;
typedef struct {
	char		*name;
	XFontStruct	*info;
} FontList;
FontList	*font_list;


usage()
{
	fprintf (stderr,"usage:  %s [-options] [-fn pattern]\n", program_name);
	fprintf (stderr, "where options include:\n");
	fprintf (stderr,
	"    -l[l[l]]                 give long info about each font\n");
	fprintf (stderr,
	"    -m                       give character min and max bounds\n");
	fprintf (stderr,
	"    -C                       force columns\n");
	fprintf (stderr,
	"    -1                       force single column\n");
	fprintf (stderr,
	"    -u                       keep output unsorted\n");
	fprintf (stderr,
	"    -o                       use OpenFont/QueryFont instead of ListFonts\n");
	fprintf (stderr,
	"    -w width                 maximum width for multiple columns\n");
	fprintf (stderr,
	"    -n columns               number of columns if multi column\n");
	fprintf (stderr,
	"    -display displayname     X server to contact\n");
	fprintf (stderr,
	    "\n");
	exit(1);
}

main(argc, argv)
int argc;
char **argv;    
{
	int	argcnt = 0, i;

	INIT_NAME;

	/* Handle command line arguments, open display */
	Setup_Display_And_Screen(&argc, argv);
	for (argv++, argc--; argc; argv++, argc--) {
		if (argv[0][0] == '-') {
			if (argcnt > 0) usage ();
			for (i=1; argv[0][i]; i++)
				switch(argv[0][i]) {
				case 'l':
					long_list++;
					break;
				case 'm':
					min_max++;
					break;
				case 'C':
					columns = 0;
					break;
				case '1':
					columns = 1;
					break;
				case 'f':
					if (--argc <= 0) usage ();
					argcnt++;
					argv++;
					get_list(argv[0]);
					goto next;
				case 'w':
					if (--argc <= 0) usage ();
					argv++;
					max_output_line_width = atoi(argv[0]);
					goto next;
				case 'n':
					if (--argc <= 0) usage ();
					argv++;
					columns = atoi(argv[0]);
					goto next;
				case 'o':
					open_instead_of_list = True;
					break;
				case 'u':
					sort_output = False;
					break;
				default:
					usage();
					break;
				}
			if (i == 1)
				usage();
		} else {
			argcnt++;
			get_list(argv[0]);
		}
	  next: ;
	}
	if (argcnt == 0)
		get_list("*");
	show_fonts();
	exit(0);
}

get_list(pattern)
	char	*pattern;
{
	int	available = nnames+1,
		i;
	char	**fonts;
	XFontStruct	*info;

	/* Get list of fonts matching pattern */
	for (;;) {
		if (open_instead_of_list) {
		    info = XLoadQueryFont (dpy, pattern);

		    if (info) {
			fonts = &pattern;
			available = 1;
			XUnloadFont (dpy, info->fid);
		    } else {
			fonts = NULL;
		    }
		    break;
		}
		    
		if (long_list == L_MEDIUM)
			fonts = XListFontsWithInfo(dpy,
				pattern, nnames, &available, &info);
		else
			fonts = XListFonts(dpy, pattern, nnames, &available);
		if (fonts == NULL || available < nnames)
			break;
		if (long_list == L_MEDIUM)
			XFreeFontInfo(fonts, info, available);
		else
			XFreeFontNames(fonts);
		nnames = available * 2;
	}

	if (fonts == NULL) {
		fprintf(stderr, "%s: pattern \"%s\" unmatched\n",
			program_name, pattern);
		return;
	}

	font_list = (FontList *)Realloc((char *)font_list,
		(font_cnt + available) * sizeof(FontList));
	for (i=0; i<available; i++) {
		font_list[font_cnt].name = fonts[i];
		if (long_list == L_MEDIUM)
			font_list[font_cnt].info = info + i;
		else
			font_list[font_cnt].info = NULL;
		font_cnt++;
	}
}

compare(f1, f2)
	FontList	*f1, *f2;
{
	char	*p1 = f1->name,
		*p2 = f2->name;

	while (*p1 && *p2 && *p1 == *p2)
		p1++, p2++;
	return(*p1 - *p2);
}

show_fonts()
{
	int	i;

	if (font_cnt == 0)
		return;

	/* first sort the output */
	if (sort_output) qsort(font_list, font_cnt, sizeof(FontList), compare);

	if (long_list > L_MEDIUM) {
	    for (i = 0; i < font_cnt; i++) {
		do_query_font (dpy, font_list[i].name);
	    }
	    return;
	}

	if (long_list == L_MEDIUM) {
		XFontStruct	*pfi;
		char		*string;

		printf("DIR  ");
		printf("MIN  ");
		printf("MAX ");
		printf("EXIST ");
		printf("DFLT ");
		printf("PROP ");
		printf("ASC ");
		printf("DESC ");
		printf("NAME");
		printf("\n");
		for (i=0; i<font_cnt; i++) {
			pfi = font_list[i].info;
			if (!pfi) {
			    fprintf (stderr, 
		    	     "%s:  no font information for font \"%s\".\n",
				     program_name, 
				     font_list[i].name ? 
				     font_list[i].name : "");
			    continue;
			}
			switch(pfi->direction) {
			case FontLeftToRight: string = "-->"; break;
			case FontRightToLeft: string = "<--"; break;
			default:	      string = "???"; break;
			}
			printf("%-4s", string);
			if (pfi->min_byte1 == 0
			 && pfi->max_byte1 == 0) {
				printf(" %3d ", pfi->min_char_or_byte2);
				printf(" %3d ", pfi->max_char_or_byte2);
			} else {
				printf("*%3d ", pfi->min_byte1);
				printf("*%3d ", pfi->max_byte1);
			}
			printf("%5s ", pfi->all_chars_exist ? "all" : "some");
			printf("%4d ", pfi->default_char);
			printf("%4d ", pfi->n_properties);
			printf("%3d ", pfi->ascent);
			printf("%4d ", pfi->descent);
			printf("%s\n", font_list[i].name);
			if (min_max) {
				char	min[ BUFSIZ ],
					max[ BUFSIZ ];
				char	*pmax = max,
					*pmin = min;

				strcpy(pmin, "     min(l,r,w,a,d) = (");
				strcpy(pmax, "     max(l,r,w,a,d) = (");
				pmin += strlen(pmin);
				pmax += strlen(pmax);

				copy_number(&pmin, &pmax,
					pfi->min_bounds.lbearing,
					pfi->max_bounds.lbearing);
				*pmin++ = *pmax++ = ',';
				copy_number(&pmin, &pmax,
					pfi->min_bounds.rbearing,
					pfi->max_bounds.rbearing);
				*pmin++ = *pmax++ = ',';
				copy_number(&pmin, &pmax,
					pfi->min_bounds.width,
					pfi->max_bounds.width);
				*pmin++ = *pmax++ = ',';
				copy_number(&pmin, &pmax,
					pfi->min_bounds.ascent,
					pfi->max_bounds.ascent);
				*pmin++ = *pmax++ = ',';
				copy_number(&pmin, &pmax,
					pfi->min_bounds.descent,
					pfi->max_bounds.descent);
				*pmin++ = *pmax++ = ')';
				*pmin = *pmax = '\0';
				printf("%s\n", min);
				printf("%s\n", max);
			}
		}
		return;
	}

	if ((columns == 0 && isatty(1)) || columns > 1) {
		int	width,
			max_width = 0,
			lines_per_column,
			j,
			index;

		for (i=0; i<font_cnt; i++) {
			width = strlen(font_list[i].name);
			if (width > max_width)
				max_width = width;
		}
		if (max_width == 0)
			Fatal_Error("all %d fontnames listed are zero length",
				    font_cnt);

		if (columns == 0) {
		    if ((max_width * 2) + output_line_padding >
			max_output_line_width) {
			columns = 1;
		    } else {
			max_width += output_line_padding;
			columns = ((max_output_line_width +
				    output_line_padding) / max_width);
		    }
		} else {
		    max_width += output_line_padding;
		}
		if (columns <= 1) goto single_column;

		if (font_cnt < columns)
			columns = font_cnt;
		lines_per_column = (font_cnt + columns - 1) / columns;

		for (i=0; i<lines_per_column; i++) {
			for (j=0; j<columns; j++) {
				index = j * lines_per_column + i;
				if (index >= font_cnt)
					break;
				if (j+1 == columns)
					printf("%s", font_list[ index ].name);
				else
					printf("%-*s",
						max_width, 
						font_list[ index ].name);
			}
			printf("\n");
		}
		return;
	}

      single_column:
	for (i=0; i<font_cnt; i++)
		printf("%s\n", font_list[i].name);
}

max(i, j)
	int	i, j;
{
	if (i > j)
		return (i);
	return(j);
}

copy_number(pp1, pp2, n1, n2)
	char	**pp1, **pp2;
	int	n1, n2;
{
	char	*p1 = *pp1;
	char	*p2 = *pp2;
	int	w;

	sprintf(p1, "%d", n1);
	sprintf(p2, "%d", n2);
	w = max(strlen(p1), strlen(p2));
	sprintf(p1, "%*d", w, n1);
	sprintf(p2, "%*d", w, n2);
	p1 += strlen(p1);
	p2 += strlen(p2);
	*pp1 = p1;
	*pp2 = p2;
}



/* ARGSUSED */
static int IgnoreError(disp, event)
    Display *disp;
    XErrorEvent *event;
{
    return 0;
}

static char *bounds_metrics_title =
                      "width left  right  asc  desc   attr   keysym\n";
static char *char_metrics_fmt = 
"\t0x%02x%02x (%u)\t%4d  %4d  %4d  %4d  %4d  0x%04x  %s\n";
static char *bounds_metrics_fmt =
"\t%3s\t\t%4d  %4d  %4d  %4d  %4d  0x%04x\n";


#define PrintBounds(_what,_ptr) \
{   register XCharStruct *p = (_ptr); \
    printf (bounds_metrics_fmt, _what, p->width, p->lbearing, \
	  p->rbearing, p->ascent, p->descent, p->attributes); }


static char* stringValued [] = { /* values are atoms */
    "FAMILY_NAME",
    "FOUNDRY",
    "STYLE",
    "MONOSPACED",
    "RELATIVE_WEIGHT",
    "RELATIVE_SET",
    "CLASSIFICATION",
    "CHARSET_REGISTRY",
    "CHARSET_ENCODING",
    "QUALITY",
    "CHARSET_COLLECTIONS",
    "FULL_NAME",
    "COPYRIGHT",
    "WEIGHT_NAME",
    "SETWIDTH_NAME",
    "SLANT",
    "SPACING",
    "ADD_STYLE_NAME",
    "FONTNAME_REGISTRY",
    "FONT",
    "DEVICE_FONT_NAME",
     NULL
    };

static void PrintProperty (prop)
    XFontProp *prop;
{
    char *atom, *value;
    char nosuch[40];
    int i;
    int (*oldhandler)() = XSetErrorHandler (IgnoreError);

    atom = XGetAtomName(dpy, prop->name);
    if (!atom) {
	atom = nosuch;
	nosuch[0] = '\0';
	(void)sprintf (atom, "No such atom = %d", prop->name);
    }
    printf ("      %s", atom);
    for (i = strlen(atom); i < 22; i++) printf (" ");
    for (i = 0; ; i++) {
	if (stringValued[i] == NULL) {
	    printf ("%d\n", prop->card32);
	    break;
	}
	if (strcmp(stringValued[i], atom) == 0) {
	    value = XGetAtomName(dpy, prop->card32);
	    if (value == NULL)
		printf ("%d (expected string value)\n", prop->card32);
	    else {
		printf ("%s\n", value);
		XFree (value);
	    }
	    break;
	}
    } 
    if (atom != nosuch) XFree (atom);
    XSetErrorHandler (oldhandler);
}


ComputeFontType (fs)
    XFontStruct *fs;
{
    int i;
    Bool char_cell = True;
    char *reason = NULL;
    XCharStruct *cs;
    Atom awatom = XInternAtom (dpy, "AVERAGE_WIDTH", False);

    printf ("  font type:\t\t");
    if (fs->min_bounds.width != fs->max_bounds.width) {
	printf ("Proportional (min and max widths not equal)\n");
	return;
    }

    if (awatom) {
	for (i = 0; i < fs->n_properties; i++) {
	    if (fs->properties[i].name == awatom &&
		(fs->max_bounds.width * 10) != fs->properties[i].card32) {
		char_cell = False;
		reason = "font width not equal to AVERAGE_WIDTH";
		break;
	    }
	}
    }

    if (fs->per_char) {
	for (i = fs->min_char_or_byte2, cs = fs->per_char;
	     i <= fs->max_char_or_byte2; i++, cs++) {
	    if (cs->width == 0) continue;
	    if (cs->width != fs->max_bounds.width) {
		/* this shouldn't happen since we checked above */
		printf ("Proportional (characters not all the same width)\n");
		return;
	    }
	    if (char_cell) {
		if (cs->width < 0) {
		    if (!(cs->width <= cs->lbearing &&
			  cs->lbearing <= cs->rbearing &&
			  cs->rbearing <= 0)) {
			char_cell = False;
			reason = "ink outside bounding box";
		    }
		} else {
		    if (!(0 <= cs->lbearing &&
			  cs->lbearing <= cs->rbearing &&
			  cs->rbearing <= cs->width)) {
			char_cell = False;
			reason = "ink outside bounding box";
		    }
		}
		if (!(cs->ascent <= fs->ascent &&
		      cs->descent <= fs->descent)) {
		    char_cell  = False;
		    reason = "characters not all same ascent or descent";
		}
	    }
	}
    }

    printf ("%s", char_cell ? "Character Cell" : "Monospaced");
    if (reason) printf (" (%s)", reason);
    printf ("\n");
	
    return;
}


print_character_metrics (info)
    register XFontStruct *info;
{
    register XCharStruct *pc = info->per_char;
    register int i, j;
    unsigned n, saven;

    printf ("  character metrics:\n");
    saven = ((info->min_byte1 << 8) | info->min_char_or_byte2);
    for (j = info->min_byte1; j <= info->max_byte1; j++) {
	n = saven;
	for (i = info->min_char_or_byte2; i <= info->max_char_or_byte2; i++) {
	    char *s = XKeysymToString ((KeySym) n);
	    printf (char_metrics_fmt, j, i, n, pc->width, pc->lbearing,
		    pc->rbearing, pc->ascent, pc->descent, pc->attributes,
		    s ? s : ".");
	    pc++;
	    n++;
	}
	saven += 256;
    }
}


do_query_font (dpy, name)
    Display *dpy;
    char *name;
{
    register int i;
    register XFontStruct *info = XLoadQueryFont (dpy, name);

    if (!info) {
	fprintf (stderr, "%s:  unable to get info about font \"%s\"\n",
		 program_name, name);
	return;
    }
    printf ("name:  %s\n", name ? name : "(nil)");
    printf ("  direction:\t\t%s\n", ((info->direction == FontLeftToRight)
				     ? "left to right" : "right to left"));
    printf ("  indexing:\t\t%s\n", 
	    ((info->min_byte1 == 0 && info->max_byte1 == 0) ? "linear" :
	     "matrix"));
    printf ("  rows:\t\t\t0x%02x thru 0x%02x (%d thru %d)\n",
	    info->min_byte1, info->max_byte1,
	    info->min_byte1, info->max_byte1);
    printf ("  columns:\t\t0x%02x thru 0x%02x (%d thru %d)\n",
	    info->min_char_or_byte2, info->max_char_or_byte2,
	    info->min_char_or_byte2, info->max_char_or_byte2);
    printf ("  all chars exist:\t%s\n",
	(info->all_chars_exist) ? "yes" : "no");
    printf ("  default char:\t\t0x%04x (%d)\n",
	    info->default_char, info->default_char);
    printf ("  ascent:\t\t%d\n", info->ascent);
    printf ("  descent:\t\t%d\n", info->descent);
    ComputeFontType (info);
    printf ("  bounds:\t\t%s", bounds_metrics_title);
    PrintBounds ("min", &info->min_bounds);
    PrintBounds ("max", &info->max_bounds);
    if (info->per_char && long_list >= L_VERYLONG) 
      print_character_metrics (info);
    printf ("  properties:\t\t%d\n", info->n_properties);
    for (i = 0; i < info->n_properties; i++)
	PrintProperty (&info->properties[i]);
    printf ("\n");

    XFreeFontInfo (NULL, info, 1);
}
